### ISSUE TITLE
Other Managed Packages being stopped by dlrs Trigger

### ISSUE THREAD

**Issue by ram-jampana** [afawcett/declarative-lookup-rollup-summaries#203]
Hi Andrew,

Thank you for this wonderful App... 

We have been using DLRS for quite some time now and never any issues. 
Recently we have upgraded one of our Unmanaged Packages - "Action plans" to the Managed Version, soon after the upgrade we are receiving an Error message with an Email Alert mentioning the declarative roll up summary tasks trigger in the Error (Attached the Images). Is there anything that You can suggest me to change or fix in order for this to work ?

I have done the following testing...
1. The Action Plans package is being used to create a bundle of Tasks (preset from a Task Template) for each Opportunity.
2. We have created about 10 roll ups on the Task Object to roll up to a Custom Object.
3. If I deactivate the Roll up summary jobs on the Task Object than the Action Plans package is working fine.

Thanks Again... @ Ram

Snapshots of the Rollups on the Task:

![image](https://cloud.githubusercontent.com/assets/11051925/8563462/fa149b94-2582-11e5-8f1b-03ca3f6af9a3.png)

Snapshots of Action Plans package:

![image](https://cloud.githubusercontent.com/assets/11051925/8563469/0ea249ee-2583-11e5-80b0-fa8a96af4eb6.png)

![image](https://cloud.githubusercontent.com/assets/11051925/8563492/95fcb37a-2583-11e5-9ea1-a1b9b7c90c9c.png)

Errors Snapshots:
![image](https://cloud.githubusercontent.com/assets/11051925/8563363/791804d2-2581-11e5-8471-986bcb3dd84c.png)

![image](https://cloud.githubusercontent.com/assets/11051925/8563369/8e7fd566-2581-11e5-95ac-a4274c3bcc31.png)

**Comment by afawcett** [comment]
Wow thats a lot of queries, i wonder if there is some recursion going on? Can you attach a debug log?

**Comment by afawcett** [comment]
Also note that your using criteria on each rollup, which will result in 14 queries per tool invocation (sorry no platform way to have multiple criteria in aggregate SOQL queries). Which is still within the 100 limit, however if there is some non-bulkified code in Action Plans and/or its doing recursion for some reason this could go up. One option to resolve this would be to switch to schedule mode, though before this i'm still keen to see a debug log. It's also odd this didn't happen with the unmanaged version, was it the same code base?

**Comment by ram-jampana** [comment]
Hi Andrew,

I will check the Schedule mode and will update you the results. Yes, the Managed and Unmanaged versions were same code base for Action plans App... 

I believe I cannot add files here in the comments in order to send the Debug log.

I have sent an Email with the Debug log to Andrew Fawcett <notifications@github.com>

Please let me know if you want me to mail it to a different Email address.

Regards,
Ram J

**Comment by afawcett** [comment]
Can you give me your org ID, i'll grab your email address and send you an email.

**Comment by ram-jampana** [comment]
Hi Andrew,

Our Org ID: 00D90000000yYEf

Just in case you need access to our Salesforce instance I have already
granted the login access.
My Email Address ram.j@guidedogsvictoria.com.au


Please let me know if you want me to do anything from my end..

Thanks,
Ram J

On 9 July 2015 at 08:50, Andrew Fawcett <notifications@github.com> wrote:

> Can you give me your org ID, i'll grab your email address and send you an
> email.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/afawcett/declarative-lookup-rollup-summaries/issues/203#issuecomment-119752960>
> .
>



-- 
Regards,

Ram Jampana
0418-991-716

**Comment by afawcett** [comment]
Thanks, just sent you an email.

**Comment by afawcett** [comment]
I've run this debug log through a nice visual tool for exploring debug logs [here](http://htmlpreview.github.io/?https://raw.github.com/codehutch/sIndent/master/index.html), using the Fancy view. Simple past in your debug log and click the Fancy button to explore easily.

You can see that the Action Plans VF page makes 9 separate calls to the [upsert] DML operation, as apposed to Apex development best practices to make just one bulkified call. This lack of [bulkification](https://developer.salesforce.com/page/Best_Practice%3A_Bulkify_Your_Code) causes the DLRS trigger to run at least twice (only counting after insert/update trigger invocations) for each of these upsert calls, since an upsert can insert new records and update others, which appears to be the case here as well. The result is the DLRS trigger code running at least 18 times, which is far more invocations than is expected. 

<img width="1078" alt="screen shot 2015-07-21 at 20 58 29" src="https://cloud.githubusercontent.com/assets/1167760/8811120/909b86e6-2feb-11e5-9b7a-4dc2ac5ba451.png">

In some cases an upsert from the page causes the DLRS trigger to run more than twice, this can only be due to Workflow rules, as per the Trigger Invocation process [here](https://developer.salesforce.com/docs/atlas.en-us.apexcode.meta/apexcode/apex_triggers_order_of_execution.htm).

<img width="735" alt="screen shot 2015-07-21 at 21 08 08" src="https://cloud.githubusercontent.com/assets/1167760/8811256/cc09e6a4-2fec-11e5-8c4f-33d01bbefec0.png">

I am not sure why this is not occuring in the unmanaged version tbh? I do see some Conga stuff in the debug log as well, was this present during the unmanaged code being active. It would be interesting to see what a debug log looked like when running the unmanaged code but i guess thats not possible now.

The general summary is i would share this post with the Action View developers and see what they think or see if you can get a developer to look at the code (via the unmanaged version) in the Action View VF page Apex controller to see if you can see **upsert** calls being made in a loop, again contray to Apex bulkification best practices.

In short i'm afraid this is nothing i can help with if the calling code is not bulkified itself. Hope this analysis helps you progress this issue and sorry its taken so long to respond.

**Comment by ram-jampana** [comment]
Hi Andrew,

Thank you for the update and your patience. I will check to verify the
Unmanaged version for the Code difference and will also check for other
options in the mean time.

Your work is greatly appreciated.... Thanks again.

Regards,
Ram Jampana

On 22 July 2015 at 06:14, Andrew Fawcett <notifications@github.com> wrote:

> I've run this debug log through a nice visual tool for exploring debug
> logs here
> <http://htmlpreview.github.io/?https://raw.github.com/codehutch/sIndent/master/index.html>,
> using the Fancy view. Simple past in your debug log and click the Fancy
> button to explore easily.
>
> You can see that the Action Plans VF page makes 9 separate calls to the
> [upsert] DML operation, as apposed to Apex development best practices to
> make just one bulkified call. This lack of bulkification
> <https://developer.salesforce.com/page/Best_Practice%3A_Bulkify_Your_Code>
> causes the DLRS trigger to run at least twice (only counting after
> insert/update trigger invocations) for each of these upsert calls, since an
> upsert can insert new records and update others, which appears to be the
> case here as well. The result is the DLRS trigger code running at least 18
> times, which is far more invocations than is expected.
>
> [image: screen shot 2015-07-21 at 20 58 29]
> <https://cloud.githubusercontent.com/assets/1167760/8811120/909b86e6-2feb-11e5-9b7a-4dc2ac5ba451.png>
>
> In some cases an upsert from the page causes the DLRS trigger to run more
> than twice, this can only be due to Workflow rules, as per the Trigger
> Invocation process here
> <https://developer.salesforce.com/docs/atlas.en-us.apexcode.meta/apexcode/apex_triggers_order_of_execution.htm>
> .
>
> [image: screen shot 2015-07-21 at 21 08 08]
> <https://cloud.githubusercontent.com/assets/1167760/8811256/cc09e6a4-2fec-11e5-8c4f-33d01bbefec0.png>
>
> I am not sure why this is not occuring in the unmanaged version tbh? I do
> see some Conga stuff in the debug log as well, was this present during the
> unmanaged code being active. It would be interesting to see what a debug
> log looked like when running the unmanaged code but i guess thats not
> possible now.
>
> The general summary is i would share this post with the Action View
> developers and see what they think or see if you can get a developer to
> look at the code (via the unmanaged version) in the Action View VF page
> Apex controller to see if you can see *upsert* calls being made in a
> loop, again contray to Apex bulkification best practices.
>
> In short i'm afraid this is nothing i can help with if the calling code is
> not bulkified itself. Hope this analysis helps you progress this issue and
> sorry its taken so long to respond.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/afawcett/declarative-lookup-rollup-summaries/issues/203#issuecomment-123464681>
> .
>



-- 
Regards,

Ram Jampana
0418-991-716

**Comment by ram-jampana** [comment]
Hi Andrew,

I have a quick question about the scheduling of the Declarative Rollup
Summary tool, I have changed the DLRS jobs from Realtime to Scheduled and
the Action Plans package seemed to be working fine.

So, is there anyway that i can make the schedule run on a hourly basis
rather than Once a Day standard Salesforce scheduling ??

I have searched for solutions and everyone suggested creating a global
class that will invoke the "RollupJob" class on a hourly basis, but I am
unable to do that because of the Managed package class.

Can you please suggest any work around or point me in right direction...

Greatly Appreciate your help and Thank you.

Regards,
Ram Jampana


On 22 July 2015 at 09:34, Ramanjaneyulu Jampana <ram.jampana@gmail.com>
wrote:

> Hi Andrew,
>
> Thank you for the update and your patience. I will check to verify the
> Unmanaged version for the Code difference and will also check for other
> options in the mean time.
>
> Your work is greatly appreciated.... Thanks again.
>
> Regards,
> Ram Jampana
>
> On 22 July 2015 at 06:14, Andrew Fawcett <notifications@github.com> wrote:
>
>> I've run this debug log through a nice visual tool for exploring debug
>> logs here
>> <http://htmlpreview.github.io/?https://raw.github.com/codehutch/sIndent/master/index.html>,
>> using the Fancy view. Simple past in your debug log and click the Fancy
>> button to explore easily.
>>
>> You can see that the Action Plans VF page makes 9 separate calls to the
>> [upsert] DML operation, as apposed to Apex development best practices to
>> make just one bulkified call. This lack of bulkification
>> <https://developer.salesforce.com/page/Best_Practice%3A_Bulkify_Your_Code>
>> causes the DLRS trigger to run at least twice (only counting after
>> insert/update trigger invocations) for each of these upsert calls, since an
>> upsert can insert new records and update others, which appears to be the
>> case here as well. The result is the DLRS trigger code running at least 18
>> times, which is far more invocations than is expected.
>>
>> [image: screen shot 2015-07-21 at 20 58 29]
>> <https://cloud.githubusercontent.com/assets/1167760/8811120/909b86e6-2feb-11e5-9b7a-4dc2ac5ba451.png>
>>
>> In some cases an upsert from the page causes the DLRS trigger to run more
>> than twice, this can only be due to Workflow rules, as per the Trigger
>> Invocation process here
>> <https://developer.salesforce.com/docs/atlas.en-us.apexcode.meta/apexcode/apex_triggers_order_of_execution.htm>
>> .
>>
>> [image: screen shot 2015-07-21 at 21 08 08]
>> <https://cloud.githubusercontent.com/assets/1167760/8811256/cc09e6a4-2fec-11e5-8c4f-33d01bbefec0.png>
>>
>> I am not sure why this is not occuring in the unmanaged version tbh? I do
>> see some Conga stuff in the debug log as well, was this present during the
>> unmanaged code being active. It would be interesting to see what a debug
>> log looked like when running the unmanaged code but i guess thats not
>> possible now.
>>
>> The general summary is i would share this post with the Action View
>> developers and see what they think or see if you can get a developer to
>> look at the code (via the unmanaged version) in the Action View VF page
>> Apex controller to see if you can see *upsert* calls being made in a
>> loop, again contray to Apex bulkification best practices.
>>
>> In short i'm afraid this is nothing i can help with if the calling code
>> is not bulkified itself. Hope this analysis helps you progress this issue
>> and sorry its taken so long to respond.
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/afawcett/declarative-lookup-rollup-summaries/issues/203#issuecomment-123464681>
>> .
>>
>
>
>
> --
> Regards,
>
> Ram Jampana
> 0418-991-716
>



-- 
Regards,

Ram Jampana
0418-991-716

**Comment by afawcett** [comment]
Yeah, they differs the processing to background, it will likely still result in a lot of calls to DLRS triggers, but the processing in the trigger is much cheaper in schedule mode (all it does is log the parent record needs a recalc, which it will just do over and over in your case). As regards running hourly you need to run a one liner from Developer Console. 

Take a look at [this](http://salesforce.stackexchange.com/questions/16337/schedule-hourly-batch-using-cron-expression-start-on-abritrary-day-of-month). Something like this will do it....

    System.schedule('Hourly', '0 0 * * * ?', new dlrs. RollupJob() );

**Comment by afawcett** [comment]
Please make sure you have the latest version installed btw, there was an important fix to scheduled mode recently, if your using child records with polymorphic lookups to multiple parent objects, not sure if thats your case, but worth being on the latest anyway.

**Comment by ram-jampana** [comment]
Hi Andrew,

Thank you for the Update.

I came across the same blog that you have suggested and executed the class
to run every hour in our Sandbox instance yesterday which seems to be
working as it should.

When I looked at the Apex Jobs list today it says that it ran 1 batch every
time it is executed but there are no new records to be processed by this
class (sandbox has No new Data put in to process) - So is this what you are
suggesting (which it will just do over and over in your case).

I have upgraded the package to the recent version.

Thank you Andrew..

Regards,
Ram

On 24 July 2015 at 10:51, Andrew Fawcett <notifications@github.com> wrote:

> Please make sure you have the latest version installed btw, there was an
> important fix to scheduled mode recently, if your using child records with
> polymorphic lookups to multiple parent objects, not sure if thats your
> case, but worth being on the latest anyway.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/afawcett/declarative-lookup-rollup-summaries/issues/203#issuecomment-124282399>
> .
>



-- 
Regards,

Ram Jampana
0418-991-716

**Comment by afawcett** [comment]
Yes the job will only process records if users have changed child records. There is a Scheduled Items object in the tool that records parent records that need updating when the schedule next runs. Once they are processed the scheduled items are deleted. Scheduled items are added as users update or insert child records.
